File: JavaSolution/src/main/java/problems/P07ThousandFirstPrime.java
package main.java.problems;

import java.util.ArrayList;
import java.util.List;

/**
 * By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
 * What is the 10,001st prime number?
 */
public class P07ThousandFirstPrime {
    private static P07ThousandFirstPrime problem1 = null;

    protected P07ThousandFirstPrime() {
        // no instantiation
    }

    public static void solveProblem() {
        System.out.println("Problem 7: 10,001st Prime");

        List<Long> primes = new ArrayList<>();
        primes.add(2L);
        primes.add(3L);
        Long possiblePrime = primes.get(primes.size() - 1) + 2;

        while (primes.size() < 10001) {
            Boolean isPrime = true;
                for (int i = 0; i <= primes.size(); i++) {
                    Long prime = primes.get(i);
                    if (prime > Math.sqrt(possiblePrime)) break;
                    else if (possiblePrime % prime == 0) {
                        isPrime = false;
                        break;
                    }
                }
            if (isPrime) primes.add(possiblePrime);
            possiblePrime += 2;
        }

        System.out.println(primes.get(primes.size() - 1));
    }
}

File: ScalaSolution/src/main/scala/problems/P07ThousandFirstPrime.scala
package main.scala.problems

import scala.collection.mutable
import scala.util.control.Breaks._

/**
  * By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
  * What is the 10,001st prime number?
  */
object P07ThousandFirstPrime {
  println("Problem 7: 10,001st Prime")

  var primes: mutable.ListBuffer[Long] = mutable.ListBuffer.empty
  primes += 2
  primes += 3
  var possiblePrime: Long = primes.last + 2

  while (primes.size < 10001) {
    var isPrime: Boolean = true
    breakable {
      for (i <- primes) {
        if (i > Math.sqrt(possiblePrime)) break
        else if (possiblePrime % i == 0) {
          isPrime = false
          break
        }
      }
    }
    if (isPrime) primes += possiblePrime
    possiblePrime += 2
  }

  println (primes.last)

}
